openedx/core/djangoapps/call_stack_manager/tests/tests.py

```diff
@@ -0,0 +1,162 @@
+
+    def test_queryset(self):
+        """ Tests for Overriding QuerySet API
+        1. Tests if classes with CallStackManager gets logged.
+        """
+        with LogCapture() as l:
+            # create and save objects of class not overriding queryset API
+            modelmixin_obj3 = ModelMixin(id_field=1, float_field=12.89)
+            modelmixin_obj4 = ModelMixin(id_field=1, float_field=23.56)
+            modelmixin_obj3.save()
+            modelmixin_obj4.save()
+
+            modelmixinCSM_obj1 = ModelMixinCSM(id_field=1, string_field="Thou shall not pass")
+            modelmixinCSM_obj2 = ModelMixinCSM(id_field=1, string_field="Not all those who wonder are lost")
+            modelmixinCSM_obj1.save()
+            modelmixinCSM_obj2.save()
+
+            ModelMixinCSM.objects.all()
+
+            # class not using Manager, should not get logged
+            ModelMixin.objects.all()
+
+            # Example - logging new call stack for openedx.core.djangoapps.call_stack_manager.tests.models.modelmixin
+            latest_log = l.records[-1].getMessage()[:l.records[-1].getMessage().find(':')]
+            latest_class = latest_log[latest_log.rfind('.') + 1:]
+
+            # desired latest class here
+            desired_class = "ModelMixinCSM"
+
+            self.assertEqual(latest_class, desired_class, msg="Latest logged event should belong to " + desired_class)
+
+    def test_donottrack(self):
+        """ Test for @donottrack
+        1. calls in decorated function should not get tracked
+        """
+        with LogCapture() as l:
+            donottrack_func()
+            self.assertEqual(len(l.records), 0, msg="Check @donottrack. should not log anything here!")
+
+    def test_parameterized_donottrack(self):
+        """ Test for parameterized @donottrack
+        1. Should not log calls of classes specified in the decorator @donotrack
+        """
+        with LogCapture() as l:
+            donottrack_parent_func()
+
+            # Example - logging new call stack for openedx.core.djangoapps.call_stack_manager.tests.models.modelmixin
+            latest_log = l.records[-1].getMessage()[:l.records[-1].getMessage().find(':')]
+            latest_class = latest_log[latest_log.rfind('.') + 1:]
+
+            # desired latest class here
+            desired_class = "ModelMixinCSM"
+
+            self.assertEqual(latest_class, desired_class,
+                             msg="The latest log should be of the class" + desired_class + "not" + latest_class)
+
+    def test_nested_parameterized_donottrack(self):
+        """ Tests parameterized nested @donottrack
+        1. should not track call of classes specified in decorated with scope bounded to the respective class
+        """
+        with LogCapture() as l:
+            #  class with CallStackManager as Manager
+            modelAnotherCSM_obj = ModelAnotherCSM(id_field=1, name_field="The Crownless again shall be king")
+            modelAnotherCSM_obj.save()
```

Comment:
Good call testing that `donottrack` doesn't take effect before the function that it's decorating.

It might be worth asserting that it's turned off again after the function finishes.